qemu/qemu.go

```diff
@@ -1275,90 +1275,149 @@ func (vhostuserDev VhostUserDevice) Valid() bool {
-		qemuParams = append(qemuParams, "-netdev")
-		qemuParams = append(qemuParams, strings.Join(netParams, ","))
+	return qemuParams
+}
+
+// QemuBlkParams builds QEMU device parameters for a VhostUserBlk device
+func (vhostuserDev VhostUserDevice) QemuBlkParams(config *Config) []string {
+	var qemuParams []string
+	var devParams []string
+
+	driver := vhostuserDev.deviceName(config)
+	if driver == "" {
+		return nil
 	}
+
+	devParams = append(devParams, driver)
+	devParams = append(devParams, "logical_block_size=4096")
+	devParams = append(devParams, "size=512M")
+	devParams = append(devParams, fmt.Sprintf("chardev=%s", vhostuserDev.CharDevID))
+
+	if vhostuserDev.Transport.isVirtioPCI(config) && vhostuserDev.ROMFile != "" {
+		devParams = append(devParams, fmt.Sprintf("romfile=%s", vhostuserDev.ROMFile))
+	}
+
+	qemuParams = append(qemuParams, "-device")
+	qemuParams = append(qemuParams, strings.Join(devParams, ","))
+
+	return qemuParams
+}
+
+// QemuFSParams builds QEMU device parameters for a VhostUserFS device
+func (vhostuserDev VhostUserDevice) QemuFSParams(config *Config) []string {
+	var qemuParams []string
+	var devParams []string
+
+	driver := vhostuserDev.deviceName(config)
+	if driver == "" {
+		return nil
+	}
+
+	devParams = append(devParams, driver)
+	devParams = append(devParams, fmt.Sprintf("chardev=%s", vhostuserDev.CharDevID))
+	devParams = append(devParams, fmt.Sprintf("tag=%s", vhostuserDev.Tag))
+	if vhostuserDev.CacheSize != 0 {
+		devParams = append(devParams, fmt.Sprintf("cache-size=%dM", vhostuserDev.CacheSize))
+	}
+	if vhostuserDev.SharedVersions {
+		devParams = append(devParams, "versiontable=/dev/shm/fuse_shared_versions")
+	}
+	if vhostuserDev.Transport.isVirtioPCI(config) && vhostuserDev.ROMFile != "" {
+		devParams = append(devParams, fmt.Sprintf("romfile=%s", vhostuserDev.ROMFile))
+	}
+
 	qemuParams = append(qemuParams, "-device")
 	qemuParams = append(qemuParams, strings.Join(devParams, ","))
 
 	return qemuParams
 }
 
+// QemuParams returns the qemu parameters built out of this vhostuser device.
+func (vhostuserDev VhostUserDevice) QemuParams(config *Config) []string {
+	var qemuParams []string
+	var charParams []string
+	var devParams []string
+
+	charParams = append(charParams, "socket")
+	charParams = append(charParams, fmt.Sprintf("id=%s", vhostuserDev.CharDevID))
+	charParams = append(charParams, fmt.Sprintf("path=%s", vhostuserDev.SocketPath))
+
+	qemuParams = append(qemuParams, "-chardev")
+	qemuParams = append(qemuParams, strings.Join(charParams, ","))
+
+	switch vhostuserDev.VhostUserType {
+	case VhostUserNet:
+		devParams = vhostuserDev.QemuNetParams(config)
+	case VhostUserSCSI:
+		devParams = vhostuserDev.QemuSCSIParams(config)
+	case VhostUserBlk:
+		devParams = vhostuserDev.QemuBlkParams(config)
+	case VhostUserFS:
+		devParams = vhostuserDev.QemuFSParams(config)
+	}
+
+	if devParams != nil {
+		qemuParams = append(qemuParams, devParams...)
+	} else {
+		return nil
+	}
+
+	return qemuParams
```

Comment:
This should be enough, shouldn't it?

```go
	return append(qemuParams, devParams...)
```

Reply:
@devimc, nops, the append is done a few lines above if devParams is not nil.

Reply:
:thinking:  but we can simplify that expression (if..else are not needed), right @Jakob-Naucke  ?

Reply:
Huh, I thought I had commented already. Yes, we need to check, because `QemuWhateverParams` might return `nil` and then we should return `nil` altogether instead of the chardev.
We can go without an `else` if that's what you mean (see update), but I think we need `if`.

Reply:
okay this looks much better :smile:  thanks